### ISSUE TITLE
[Doc]: add sphinx versioning directives

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#23506]
### Problem

It can be unclear which features were added when just by looking at the docs, and would be helpful if at at least major features (and whatever else folks think would be a good idea) could be flagged with a version added/changed/etc directive. 

### Suggested improvement

Introduce the [versioning directives](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded)

Per the call, a first step is to 
- [ ] create a PR with directives added to some of the 3.6 PRS

Additionally
- [ ] sort out guidance for adding directive (to which prs) and update PR guidelines accordingly

**Comment by tacaswell** [comment]
Before we start adding the directive to any docstrings, we should document a policy on it and get that merged.

**Comment by timhoffm** [comment]
### Policy proposal:

`.. versionadded::` should be added to the docstring for new classes, functions, methods and parameters. The entry should be placed at the end of the description block and before the Parameters section. For parameters, it should be placed at the end of the description. For minor releases 3.x.0 the bugfix-version should be left out i.e. `.. versionadded:: 3.x`.

```
class Foo:
    """
    This is the summary.

    Followed by a longer description block. Consisting of multiple lines
    and paragraphs.

    .. versionadded:: 3.5

    Parameters
    ----------
    a : int
        The first parameter.
    b: bool, default: False
        This was added later.

        .. versionadded:: 3.6
    """

    def set_b(b):
        """
        Set b.

        .. versionadded:: 3.6

        Parameters
        ----------
        b: bool
```
        
### Note
While supported I would not add versionadded to whole modules. Instead, let's simply tag all contained elements. That's more discoverable.

**Comment by mwaskom** [comment]
This would be a very welcome addition for downstream libraries.

For documentation of specific parameters, the admonition produced by versionadded directive takes up a lot of vertical space, so I wonder whether just having "Put `"Added in version x"` in the description is a preferable policy.

**Comment by story645** [comment]
@melissawm had the (great) suggestion that this should be addressed as two concurrent PRs - one with the policy and one with some sample output. Doing so would allow for easier policy tweaking on the basis of output while still keeping them independent.

@timhoffm my thinking was: 

- if it gets a what's new, it gets a `version added`
- if it gets an API change, it gets a `versionchanged`

mostly because that seemed consistent with current policy and very easy to document. I don't think your comment contradicts and is more guidance on placement, but I'm not sure.

**Comment by timhoffm** [comment]
> For documentation of specific parameters, the admonition produced by versionadded directive takes up a lot of vertical space, so I wonder whether just having "Put "Added in version x" in the description is a preferable policy.

I belive there is value in having this styled differently from plain text. We can tune the spacing using CSS:

e.g. adding
```
dl.field-list .versionadded p {
    margin-top: 0.1em;
    margin-bottom: 0.1em;
}

dl.field-list .versionadded {
    margin-top: 0.3em;
}

quickly tested on [pandas](https://pandas.pydata.org/docs/reference/api/pandas.io.formats.style.Styler.format.html) results in 
```
<img width="470" alt="image" src="https://user-images.githubusercontent.com/2836374/182253579-7f930e1b-a910-4de7-9379-bacb46cf832a.png">

instead of 
<img width="471" alt="image" src="https://user-images.githubusercontent.com/2836374/182253639-e06c8976-89c3-4678-a609-d10d95ec0f72.png">

**Comment by timhoffm** [comment]
@story645  @melissawm I don't follow why it's important to keep two independent PRs. Policy and some first samples must be developed together and match. I would have no issue doing this simultaneously within one PR.

**Comment by story645** [comment]
Cause Tom said in this thread that policy should be [merged first](https://github.com/matplotlib/matplotlib/issues/23506#issuecomment-1198697710) but the discussion on the call was about creating some samples first.

**Comment by tacaswell** [comment]
My memory of the call was that the next action was a PR adding the policy (basically what @timhoffm wrote out in https://github.com/matplotlib/matplotlib/issues/23506#issuecomment-1200108231) as a starting place for the discussion, but I see the notes do not agree with my memory so I apologize for not making sure everyone was on the same page before we moved on.

I agree with Tim that doing the policy and a few examples (plus it seems some styling work) in the same PR makes good sense.

**Comment by jklymak** [comment]
The notes are idiosyncratic 😉. If anything in them is meant to be a formal roadmap forward for a subject, we should probably stop and write it out carefully.  

But I don't think this particular issue is particularly contentious that we can't be flexible on how to proceed.  In particular my understanding is that we wanted a policy before we expect everyone to add these going forward, but I don't think that precluded a few examples as well.

**Comment by tacaswell** [comment]
I have a concern that "some" is worse than "none" and that this may be a case were (at least going forward), we need to boil the ocean.

**Comment by story645** [comment]
>have a concern that "some" is worse than "none"

Which is one of the reasons Melissa suggested the split PRs & I think it's a good idea. Then the policy can go in and we can build up the examples (for example 1 PR per major doc version). It also lets us split up the work more cleanly.

**Comment by jklymak** [comment]
I thought the plan was to just do this going forward. If we have to go back and do every method we will never do this.

**Comment by story645** [comment]
> I thought the plan was to just do this going forward. If we have to go back and do every method we will never do this.
Yeah, my rough plan is target 3.7 for the policy, maybe 3.6 if all the 3.6 features can get marked up. Then over time maybe the previous releases get marked up, which is mostly tedious but not difficult if we can agree on a straightforward enough policy.
